stacker/util.py

```diff
@@ -531,10 +536,69 @@ def create_cache_directories(self):
 
     def get_package_sources(self):
         """Makes remote python packages available for local use."""
+        # Checkout S3 repositories specified in config
+        for config in self.sources.get('s3', []):
+            self.fetch_s3_package(config=config)
         # Checkout git repositories specified in config
         for config in self.sources.get('git', []):
             self.fetch_git_package(config=config)
 
+    def fetch_s3_package(self, config):
+        """Makes a remote S3 archive available for local use
+
+        Args:
+            config (:class:`stacker.config.GitPackageSource`): git config
+                dictionary
+
+        """
+        if config['key'].endswith('.tar.gz'):
+            archive_type = '.tar.gz'
+            dir_name = self.sanitize_uri_path(
+                "s3-%s-%s" % (config['bucket'], config['key'][:-7])
+            )
+        elif config['key'].endswith('.zip'):
+            archive_type = '.zip'
+            dir_name = self.sanitize_uri_path(
+                "s3-%s-%s" % (config['bucket'], config['key'][:-4])
+            )
+        else:
+            raise ValueError(
+                "Archive type could not be determined for S3 object \"%s\" "
+                "in bucket %s." % (config['key'], config['bucket'])
+            )
+
+        # We can skip downloading the archive if it's already been cached
+        cached_dir_path = os.path.join(self.package_cache_dir, dir_name)
```

Comment:
The caching logic here seems like it shares similarities with the git package sources - should this be factored out into a common method(s)?